[Package]: recoll

### Package description

Recoll is a full-text search tool which (currently) runs on Linux, MacOS, and MS Windows. While there is a GUI provided, it is entirely optional, and there is a [web based frontend](https://framagit.org/medoc92/recollwebui/) which can be used to run on basically any platform.
The program has already been a long time in the Debian [main repository](https://packages.debian.org/source/stable/recoll), so I presume that counts as "well-known".

### Home page URL

https://www.lesbonscomptes.com/recoll/index.html

### Source code URL

https://framagit.org/medoc92/recoll

### Packaging policy acknowledgement

- [X] I have read and understand the [Packaging Policy](https://github.com/termux/termux-packages/blob/master/CONTRIBUTING.md#packaging-policy).

### Additional information

There is a related [Recoll web UI issue](https://github.com/koniu/recoll-webui/issues/86).
I tried compiling on my device (Android 10 device on `aarch64`), but was ultimately unsuccessful due to linker errors.

### Requirements

The packages I installed using `pkg` are:

```plaintext
termux-tools
git
libtool
make
automake
clang
binutils
gettext
file
libxapian
libxslt
pkg-config
```
### Compilation

I initially cloned the repo:
```sh
git clone https://framagit.org/medoc92/recoll
```
then `cd`ed to the `src/` subdirectory, and ran:
```sh
sh autogen.sh
# disables support for CHM files, for which there doesn't appear to be a suitable library on Termux repos
# also disables any GUI support since we don't need it for the web frontend
./configure --prefix="${PREFIX}" --disable-python-chm --disable-x11mon --disable-qtgui
```
After a while, I run into a cryptic error:
```plaintext
./configure: 19872: Syntax error: word unexpected (expecting ")")
```
with the offending line being:
```plaintext
PKG_CHECK_MODULES(XSLT, libxslt, , as_fn_error $? "libxslt" "$LINENO" 5)
```
After removing this line and hoping for the best, the configuration ran successfully.
Finally, I just ran `make`, but encountered an error at the following command:
```plaintext
libtool: link: g++  -fPIC -DPIC -shared -nostdlib /data/data/com.termux/files/usr/lib/crtbegin_so.o  aspell/.libs/rclaspell.o bincimapmime/.libs/convert.o bincimapmime/.libs/mime-parsefull.o bincimapmime/.libs/mime-parseonlyheader.o bincimapmime/.libs/mime-printbody.o bincimapmime/.libs/mime.o common/.libs/webstore.o common/.libs/cstr.o common/.libs/rclconfig.o common/.libs/rclinit.o common/.libs/syngroups.o common/.libs/textsplit.o common/.libs/textsplitko.o common/.libs/unacpp.o common/.libs/utf8fn.o index/.libs/webqueuefetcher.o index/.libs/checkretryfailed.o index/.libs/exefetcher.o index/.libs/fetcher.o index/.libs/fsfetcher.o index/.libs/idxdiags.o index/.libs/idxstatus.o index/.libs/mimetype.o index/.libs/subtreelist.o internfile/.libs/extrameta.o internfile/.libs/htmlparse.o internfile/.libs/internfile.o internfile/.libs/mh_exec.o internfile/.libs/mh_execm.o internfile/.libs/mh_html.o internfile/.libs/mh_mail.o internfile/.libs/mh_mbox.o internfile/.libs/mh_text.o internfile/.libs/mh_xslt.o internfile/.libs/mimehandler.o internfile/.libs/myhtmlparse.o internfile/.libs/txtdcode.o internfile/.libs/uncomp.o query/.libs/docseq.o query/.libs/docseqdb.o query/.libs/docseqhist.o query/.libs/dynconf.o query/.libs/filtseq.o query/.libs/plaintorich.o query/.libs/qresultstore.o query/.libs/recollq.o query/.libs/reslistpager.o query/.libs/sortseq.o query/.libs/wasaparse.o query/.libs/wasaparseaux.o rcldb/.libs/daterange.o rcldb/.libs/expansiondbs.o rcldb/.libs/rclabstract.o rcldb/.libs/rclabsfromtext.o rcldb/.libs/rcldb.o rcldb/.libs/rcldoc.o rcldb/.libs/rcldups.o rcldb/.libs/rclquery.o rcldb/.libs/rclterms.o rcldb/.libs/rclvalues.o rcldb/.libs/searchdata.o rcldb/.libs/searchdatatox.o rcldb/.libs/searchdataxml.o rcldb/.libs/stemdb.o rcldb/.libs/stoplist.o rcldb/.libs/synfamily.o unac/.libs/unac.o utils/.libs/appformime.o utils/.libs/base64.o utils/.libs/cancelcheck.o utils/.libs/chrono.o utils/.libs/circache.o utils/.libs/closefrom.o utils/.libs/cmdtalk.o utils/.libs/conftree.o utils/.libs/copyfile.o utils/.libs/cpuconf.o utils/.libs/dlib.o utils/.libs/ecrontab.o utils/.libs/execmd.o utils/.libs/fileudi.o utils/.libs/fstreewalk.o utils/.libs/hldata.o utils/.libs/idfile.o utils/.libs/listmem.o utils/.libs/log.o utils/.libs/md5.o utils/.libs/md5ut.o utils/.libs/mimeparse.o utils/.libs/miniz.o utils/.libs/netcon.o utils/.libs/pathut.o utils/.libs/pxattr.o utils/.libs/rclionice.o utils/.libs/rclutil.o utils/.libs/readfile.o utils/.libs/smallut.o utils/.libs/strmatcher.o utils/.libs/transcode.o utils/.libs/utf8iter.o utils/.libs/wipedir.o utils/.libs/x11mon.o utils/.libs/zlibut.o   -Wl,-rpath -Wl,/data/data/com.termux/files/usr/lib -L/data/data/com.termux/files/usr/lib -lxslt -lxml2 -lxapian -liconv -lz -lpthread -L/system/lib64 -lc++_shared -lm -lc -l:libunwind.a -ldl /data/data/com.termux/files/usr/lib/crtend_so.o  -g -O2 -Wl,--no-undefined   -Wl,-soname -Wl,librecoll-1.32.1.so -o .libs/librecoll-1.32.1.so
```
with the error being:
```plaintext
ld.lld: error: undefined symbol: __aarch64_ldadd8_relax
>>> referenced by memory:3439 (/data/data/com.termux/files/usr/include/c++/v1/memory:3439)
>>>               index/.libs/subtreelist.o:(subtreelist(RclConfig*, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > >&))
>>> referenced by memory:3439 (/data/data/com.termux/files/usr/include/c++/v1/memory:3439)
>>>               internfile/.libs/internfile.o:(FileInterner::tempFileForMT(TempFile&, RclConfig*, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&))
>>> referenced by memory:3439 (/data/data/com.termux/files/usr/include/c++/v1/memory:3439)
>>>               internfile/.libs/internfile.o:(FileInterner::topdocToFile(TempFile&, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, RclConfig*, Rcl::Doc const&, bool))
>>> referenced 42 more times                                                                                                                                                  ld.lld: error: undefined symbol: __aarch64_ldadd8_acq_rel
>>> referenced by memory:3450 (/data/data/com.termux/files/usr/include/c++/v1/memory:3450)
>>>               index/.libs/subtreelist.o:(subtreelist(RclConfig*, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > >&))
>>> referenced by memory:3450 (/data/data/com.termux/files/usr/include/c++/v1/memory:3450)
>>>               index/.libs/subtreelist.o:(subtreelist(RclConfig*, std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&, std::__ndk1::vector<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> >, std::__ndk1::allocator<std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > > >&))
>>> referenced by memory:3450 (/data/data/com.termux/files/usr/include/c++/v1/memory:3450)
>>>               index/.libs/subtreelist.o:(std::__ndk1::shared_ptr<Rcl::SearchData>::~shared_ptr())
>>> referenced 148 more times
clang-14: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [Makefile:1648: librecoll.la] Error 1
```
According to [this](https://github.com/rust-lang/git2-rs/issues/706) and [this](https://bugzilla.redhat.com/show_bug.cgi?id=1830472) issue, I tried running both with `CFLAGS="${CFLAGS} -mno-outline-atomics"` and `CFLAGS="${CFLAGS} -moutline-atomics"`, but neither worked (`__aarch64_ldadd8_relax` and/or `__aarch64_ldadd8_acq_rel` missing), so I'm not entirely sure what to do. If anyone can provide guidance on resolving this issues, I can try compiling again and report back.

--- a/packages/recoll/build.sh
+++ b/packages/recoll/build.sh
@@ -1,0 +1,14 @@
+TERMUX_PKG_HOMEPAGE=https://www.lesbonscomptes.com/recoll/index.html
+TERMUX_PKG_DESCRIPTION="Full-text search for your desktop"
+TERMUX_PKG_LICENSE="GPL-2.0"
+TERMUX_PKG_MAINTAINER="@termux"
+TERMUX_PKG_VERSION=1.33.1
+TERMUX_PKG_SRCURL=http://www.lesbonscomptes.com/recoll/recoll-${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_SHA256=025dec0b9745e1ddacd86ee5478f9c52b2da2e5c307f831aaa5b2c7f9d7e8db9
+TERMUX_PKG_DEPENDS="aspell, libxapian, libxslt, zlib"
+TERMUX_PKG_EXTRA_CONFIGURE_ARGS=" --disable-userdoc --disable-python-module --disable-python-chm --disable-x11mon --disable-qtgui"
+
+
+termux_step_pre_configure() {
+	LDFLAGS+=" $($CC -print-libgcc-file-name)"
+}

